File: lib/VM/JSLib/Symbol.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

//===----------------------------------------------------------------------===//
/// \file
/// ES6.0 19.4 Initialize the Symbol constructor.
//===----------------------------------------------------------------------===//
#include "JSLibInternal.h"

#include "hermes/VM/Operations.h"
#include "hermes/VM/PrimitiveBox.h"
#include "hermes/VM/Runtime.h"
#include "hermes/VM/StringBuilder.h"
#include "hermes/VM/StringPrimitive.h"

namespace hermes {
namespace vm {

void createSymbolConstructor(Runtime &runtime) {
  auto symbolPrototype = Handle<JSObject>::vmcast(&runtime.symbolPrototype);

  struct : public Locals {
    PinnedValue<SymbolID> symbolHandle;
    PinnedValue<NativeConstructor> cons;
  } lv;
  LocalsRAII lraii(runtime, &lv);

  defineSystemConstructor(
      runtime,
      Predefined::getSymbolID(Predefined::Symbol),
      symbolConstructor,
      symbolPrototype,
      0,
      lv.cons);

  defineMethod(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::predefinedFor),
      nullptr,
      symbolFor,
      1);
  defineMethod(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::keyFor),
      nullptr,
      symbolKeyFor,
      1);

  // Well-known symbols.
  DefinePropertyFlags dpf{};
  dpf.writable = 0;
  dpf.enumerable = 0;
  dpf.configurable = 0;
  dpf.setWritable = 0;
  dpf.setEnumerable = 0;
  dpf.setConfigurable = 0;
  dpf.setValue = 1;

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolHasInstance);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::hasInstance),
      lv.symbolHandle,
      dpf);
  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolIterator);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::iterator),
      lv.symbolHandle,
      dpf);
  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolAsyncIterator);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::asyncIterator),
      lv.symbolHandle,
      dpf);
  lv.symbolHandle =
      Predefined::getSymbolID(Predefined::SymbolIsConcatSpreadable);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::isConcatSpreadable),
      lv.symbolHandle,
      dpf);
  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolToPrimitive);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::toPrimitive),
      lv.symbolHandle,
      dpf);
  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolToStringTag);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::toStringTag),
      lv.symbolHandle,
      dpf);

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolMatch);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::match),
      lv.symbolHandle,
      dpf);

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolMatchAll);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::matchAll),
      lv.symbolHandle,
      dpf);

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolSearch);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::search),
      lv.symbolHandle,
      dpf);

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolReplace);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::replace),
      lv.symbolHandle,
      dpf);

  lv.symbolHandle = Predefined::getSymbolID(Predefined::SymbolSplit);
  defineProperty(
      runtime,
      lv.cons,
      Predefined::getSymbolID(Predefined::split),
      lv.symbolHandle,
      dpf);

  // Symbol.prototype.xxx methods.
  void *ctx = nullptr;
  defineAccessor(
      runtime,
      symbolPrototype,
      Predefined::getSymbolID(Predefined::description),
      ctx,
      symbolPrototypeDescriptionGetter,
      nullptr,
      false,
      true);
  defineMethod(
      runtime,
      symbolPrototype,
      Predefined::getSymbolID(Predefined::toString),
      ctx,
      symbolPrototypeToString,
      0);
  defineMethod(
      runtime,
      symbolPrototype,
      Predefined::getSymbolID(Predefined::valueOf),
      ctx,
      symbolPrototypeValueOf,
      0);

  dpf = DefinePropertyFlags::getDefaultNewPropertyFlags();
  dpf.writable = 0;
  dpf.enumerable = 0;
  defineProperty(
      runtime,
      symbolPrototype,
      Predefined::getSymbolID(Predefined::SymbolToStringTag),
      runtime.getPredefinedStringHandle(Predefined::Symbol),
      dpf);
  (void)defineMethod(
      runtime,
      symbolPrototype,
      Predefined::getSymbolID(Predefined::SymbolToPrimitive),
      Predefined::getSymbolID(Predefined::squareSymbolToPrimitive),
      nullptr,
      symbolPrototypeValueOf,
      1,
      dpf);
}

CallResult<HermesValue> symbolConstructor(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  if (args.isConstructorCall()) {
    return runtime.raiseTypeError("Symbol is not a constructor");
  }

  struct : public Locals {
    PinnedValue<StringPrimitive> descString;
  } lv;
  LocalsRAII lraii{runtime, &lv};
  if (args.getArg(0).isUndefined()) {
    // If description is undefined, the descString will eventually be "".
    lv.descString = runtime.getPredefinedString(Predefined::emptyString);
  } else {
    auto descStringRes = toString_RJS(runtime, args.getArgHandle(0));
    if (LLVM_UNLIKELY(descStringRes == ExecutionStatus::EXCEPTION)) {
      return ExecutionStatus::EXCEPTION;
    }
    lv.descString = descStringRes->get();
  }

  auto symbolRes = runtime.getIdentifierTable().createNotUniquedSymbol(
      runtime, lv.descString);
  if (LLVM_UNLIKELY(symbolRes == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }

  return HermesValue::encodeSymbolValue(*symbolRes);
}

CallResult<HermesValue> symbolFor(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  auto cr = toString_RJS(runtime, args.getArgHandle(0));
  if (LLVM_UNLIKELY(cr == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }

  struct : public Locals {
    PinnedValue<StringPrimitive> key;
  } lv;
  LocalsRAII lraii(runtime, &lv);

  lv.key = std::move(*cr);
  auto symbolRes = runtime.getSymbolRegistry().getSymbolForKey(runtime, lv.key);
  if (LLVM_UNLIKELY(symbolRes == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }
  return HermesValue::encodeSymbolValue(*symbolRes);
}

CallResult<HermesValue> symbolKeyFor(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  if (LLVM_UNLIKELY(!args.getArg(0).isSymbol())) {
    return runtime.raiseTypeError("Symbol.keyFor() requires a symbol argument");
  }

  auto sym = Handle<SymbolID>::vmcast(args.getArgHandle(0));

  if (runtime.getSymbolRegistry().hasSymbol(sym.get())) {
    return HermesValue::encodeStringValue(
        runtime.getStringPrimFromSymbolID(*sym));
  }

  return HermesValue::encodeUndefinedValue();
}

/// ES10.0 19.4.3.2 get Symbol.prototype.description
/// TODO(T79770380): make the Symbol(undefined) case spec-conformant.
CallResult<HermesValue> symbolPrototypeDescriptionGetter(
    void *,
    Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  struct : public Locals {
    PinnedValue<SymbolID> sym;
  } lv;
  LocalsRAII lraii{runtime, &lv};
  // 1. Let s be the this value.
  // 2. Let sym be ? thisSymbolValue(s).
  if (args.getThisArg().isSymbol()) {
    lv.sym = args.vmcastThis<SymbolID>().get();
  } else if (auto symHandle = args.dyncastThis<JSSymbol>()) {
    lv.sym = symHandle->getPrimitiveSymbol();
  } else {
    return runtime.raiseTypeError(
        "Symbol.prototype.description can only be called on Symbol");
  }

  // 3. Return sym.[[Description]].
  StringPrimitive *desc = runtime.getStringPrimFromSymbolID(lv.sym.get());
  return HermesValue::encodeStringValue(desc);
}

/// ES10 19.4.3.3 Symbol.prototype.toString ( )
CallResult<HermesValue> symbolPrototypeToString(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  // 1. Let sym be ? thisSymbolValue(this value).
  struct : public Locals {
    PinnedValue<SymbolID> sym;
  } lv;
  LocalsRAII lraii{runtime, &lv};
  if (args.getThisArg().isSymbol()) {
    lv.sym = args.vmcastThis<SymbolID>().get();
  } else if (auto symHandle = args.dyncastThis<JSSymbol>()) {
    lv.sym = symHandle->getPrimitiveSymbol();
  } else {
    return runtime.raiseTypeError(
        "Symbol.prototype.toString can only be called on Symbol");
  }

  // 2. Return SymbolDescriptiveString(sym).
  auto str = symbolDescriptiveString(runtime, lv.sym);
  if (LLVM_UNLIKELY(str == ExecutionStatus::EXCEPTION)) {
    return ExecutionStatus::EXCEPTION;
  }
  return str->getHermesValue();
}

/// ES10 19.4.3.4 Symbol.prototype.valueOf ( )
CallResult<HermesValue> symbolPrototypeValueOf(void *, Runtime &runtime) {
  NativeArgs args = runtime.getCurrentFrame().getNativeArgs();
  // 1. Return ? thisSymbolValue(this value).
  if (args.getThisArg().isSymbol()) {
    return args.getThisArg();
  }
  if (auto jsSymbol = args.dyncastThis<JSSymbol>()) {
    return jsSymbol->getPrimitiveSymbol().getHermesValue();
  }
  return runtime.raiseTypeError(
      "Symbol.prototype.valueOf can only be called on Symbol");
}

} // namespace vm
} // namespace hermes

File: test/hermes/symbol.js
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// RUN: %hermes -O %s | %FileCheck --match-full-lines %s
"use strict";

print('Symbol()');
// CHECK-LABEL: Symbol()
print(typeof Symbol());
// CHECK-NEXT: symbol

// Ensure that dynamically allocated descriptions are marked during GC.
var sym = Symbol(44912);
print(String(sym));
// CHECK-NEXT: Symbol(44912)
gc();
print(String(sym));
// CHECK-NEXT: Symbol(44912)

print('description');
// CHECK-LABEL: description
print(Symbol().description);
// CHECK-EMPTY:
print(Symbol('asdf').description);
// CHECK-NEXT: asdf
print(Symbol({toString: function() {return 'asdf'}}).description)
// CHECK-NEXT: asdf
print(Object(Symbol('asdf')).description)
// CHECK-NEXT: asdf
try { print(Symbol.prototype.description.call(3)); }
catch (e) { print('caught', e.name); }
// CHECK-NEXT: caught TypeError
print(Symbol.iterator.description)  // well-known Symbols
// CHECK-NEXT: Symbol.iterator

// TODO(The below cases need to return `undefined` to be fully conformant.)
// print(Symbol().description)
// print(Symbol(undefined).description)

print('toString');
// CHECK-LABEL: toString
print(Symbol().toString());
// CHECK-NEXT: Symbol()
print(Symbol(undefined).toString());
// CHECK-NEXT: Symbol()
print(Symbol('asdf').toString());
// CHECK-NEXT: Symbol(asdf)
print(Symbol({toString: function() {return 'asdf'}}).toString());
// CHECK-NEXT: Symbol(asdf)
print(Object(Symbol('asdf')).toString());
// CHECK-NEXT: Symbol(asdf)
try { print(Symbol.prototype.toString.call(3)); }
catch (e) { print('caught', e.name); }
// CHECK-NEXT: caught TypeError

print('valueOf');
// CHECK-LABEL: valueOf
var sym = Symbol();
print(sym === sym.valueOf());
// CHECK-NEXT: true
print(sym === Object(sym).valueOf());
// CHECK-NEXT: true

print('equality');
// CHECK-LABEL: equality
print(Symbol() === Symbol());
// CHECK-NEXT: false
var x = Symbol();
print(x === x);
// CHECK-NEXT: true

print('object properties');
// CHECK-LABEL: object properties
var obj = {};
var x = Symbol('asdf');
var y = Symbol('asdf');
obj[x] = 1;
obj[y] = 2;
obj['asdf'] = 3;
print(obj[x], obj[y], obj['asdf']);
// CHECK-NEXT: 1 2 3
gc();
print(obj[x], obj[y], obj['asdf']);
// CHECK-NEXT: 1 2 3
delete obj[x];
print(obj[x]);
// CHECK-NEXT: undefined
var sym = Symbol();
var obj = {};
var wrapped = {
  toString: function() {
    return sym;
  }
}
obj[sym] = 3;
print(obj[sym], obj[wrapped]);
// CHECK-NEXT: 3 3
function foo() {
  print(arguments[Symbol()]);
}
foo();
// CHECK-NEXT: undefined

print('getOwnPropertySymbols');
// CHECK-LABEL: getOwnPropertySymbols
var obj = {};
var x = Symbol('x');
var y = Symbol('y');
obj[x] = 1;
obj[y] = 2;
var syms = Object.getOwnPropertySymbols(obj);
print(syms.length, String(syms[0]), String(syms[1]));
// CHECK-NEXT: 2 Symbol(x) Symbol(y)
print(obj[syms[0]], obj[syms[1]]);
// CHECK-NEXT: 1 2

print('registry');
// CHECK-LABEL: registry
var x = Symbol.for('x');
var x2 = Symbol.for('x');
var y = Symbol.for('y');
print(x === x2);
// CHECK-NEXT: true
print(x === y);
// CHECK-NEXT: false
print(String(x), String(x2), String(y));
// CHECK-NEXT: Symbol(x) Symbol(x) Symbol(y)
print(Symbol.keyFor(x));
// CHECK-NEXT: x
print(Symbol.keyFor(x2));
// CHECK-NEXT: x
print(Symbol.keyFor(y));
// CHECK-NEXT: y
print(Symbol.keyFor(Symbol('asdf')));
// CHECK-NEXT: undefined
print(Symbol.keyFor(Symbol()));
// CHECK-NEXT: undefined
try { print(Symbol.keyFor('a')); } catch (e) { print('caught', e.name); }
// CHECK-NEXT: caught TypeError

print('well-known Symbols');
// CHECK-LABEL: well-known Symbols
print(String(Symbol.iterator));
// CHECK-NEXT: Symbol(Symbol.iterator)
print(String(Symbol.toStringTag));
// CHECK-NEXT: Symbol(Symbol.toStringTag)
print(Array.prototype[Symbol.iterator] === Array.prototype.values);
// CHECK-NEXT: true
print(Uint32Array.prototype[Symbol.iterator] === Uint32Array.prototype.values);
// CHECK-NEXT: true
print(Set.prototype[Symbol.iterator] === Set.prototype.values);
// CHECK-NEXT: true
print(Map.prototype[Symbol.iterator] === Map.prototype.entries);
// CHECK-NEXT: true
print(String(Symbol.hasInstance));
// CHECK-NEXT: Symbol(Symbol.hasInstance)
print(String(Symbol.match));
// CHECK-NEXT: Symbol(Symbol.match)
print(String(Symbol.matchAll));
// CHECK-NEXT: Symbol(Symbol.matchAll)
print(String(Symbol.search));
// CHECK-NEXT: Symbol(Symbol.search)
print(String(Symbol.replace));
// CHECK-NEXT: Symbol(Symbol.replace)
print(String(Symbol.split));
// CHECK-NEXT: Symbol(Symbol.split)
